contracts/HATVaultsRegistry.sol

```diff
@@ -0,0 +1,563 @@
+error SwapFailed();
+// Wrong amount received
+error AmountSwappedLessThanMinimum();
+// Challenge period too short
+error ChallengePeriodTooShort();
+// Challenge period too long
+error ChallengePeriodTooLong();
+// Challenge timeout period too short
+error ChallengeTimeOutPeriodTooShort();
+// Challenge timeout period too long
+error ChallengeTimeOutPeriodTooLong();
+
+/** @title Registry to deploy Hats.finance vaults and manage shared parameters
+* @author hats.finance
+* @notice Hats.finance is a proactive bounty protocol for white hat hackers
+* and auditors, where projects, community members, and stakeholders
+* incentivize protocol security and responsible disclosure.
+* Hats create scalable vaults using the project’s own token. The value of the
+* bounty increases with the success of the token and project.
+*
+* The HATVaultsRegistry defines a few roles which are relevant to all vaults:
+* Governance - The owner of HATVaultsRegistry and every HATVault created, has
+* the permission to set the feeSetter and arbitrator roles, to set time limits
+* and bounty parameters and change vaults' info.
+* Arbitrator - Can challenge submitted claims for bounty payouts, approve them
+* with a different bounty percentage or dismiss them.
+* FeeSetter - The only address which can set the fee on withdrawals on all
+* vaults.
+*
+* This project is open-source and can be found at:
+* https://github.com/hats-finance/hats-contracts
+*/
+contract HATVaultsRegistry is Ownable {
+    using SafeERC20 for IERC20;
+
+    struct GeneralParameters {
+        // vesting duration for the part of the bounty given to the hacker in HAT tokens
+        uint256 hatVestingDuration;
+        // vesting periods for the part of the bounty given to the hacker in HAT tokens
+        uint256 hatVestingPeriods;
+        // withdraw enable period. safetyPeriod starts when finished.
+        uint256 withdrawPeriod;
+        // withdraw disable period - time for the committee to gather and decide on actions,
+        // withdrawals are not possible in this time. withdrawPeriod starts when finished.
+        uint256 safetyPeriod;
+        // period of time after withdrawRequestPendingPeriod where it is possible to withdraw
+        // (after which withdrawals are not possible)
+        uint256 withdrawRequestEnablePeriod;
+        // period of time that has to pass after withdraw request until withdraw is possible
+        uint256 withdrawRequestPendingPeriod;
+        // period of time that has to pass after setting a pending max
+        // bounty before it can be set as the new max bounty
+        uint256 setMaxBountyDelay;
+        // fee in ETH to be transferred with every logging of a claim
+        uint256 claimFee;  
+    }
+
+    struct SwapData {
+        uint256 totalHackersHatReward;
+        uint256 amount;
+        uint256 amountUnused;
+        uint256 hatsReceived;
+        uint256 totalHackerReward;
+    }
+
+    address public immutable hatVaultImplementation;
```

Comment:
thinking if we want this to be settable. But probaly no need, we can always get a new registry..

Reply:
wdym?